### PR TITLE
strings: replace function with strongly typed wrapper

### PR DESCRIPTION
**Pull request by scalbatty**

### CODE REVIEW

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift [hunk 1]
```diff
@@ -295,6 +263,16 @@ extension Locale {
   }
 }
 
+private func snakeToCamel(_ string: String) -> String {
+    return string.split(separator: "_")
+        .enumerated()
+        .map {
+            if $0.0 == 0 { return String($0.1) }
+            else { return String($0.1).localizedCapitalized }
```
Comment (cyrilmottier) [inline Sources/RswiftCore/Generators/StringsStructGenerator.swift]: Clearly that's not what you want in code generation. Never use localized version of capitalization as you want always the same result regardless of your current locale.

 - User read text: `localizedCapitalized`
 - Machine read text: `capitalized`

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift [hunk 2]
```diff
@@ -295,6 +263,16 @@ extension Locale {
   }
 }
 
+private func snakeToCamel(_ string: String) -> String {
+    return string.split(separator: "_")
+        .enumerated()
+        .map {
+            if $0.0 == 0 { return String($0.1) }
```
Comment (cyrilmottier) [inline Sources/RswiftCore/Generators/StringsStructGenerator.swift]: Isn't `$0.1` already a `String`?

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift (hunk 1 above)
Comment (cyrilmottier) [reply Sources/RswiftCore/Generators/StringsStructGenerator.swift]: Isn't `$0.1` already a `String`?

File: Sources/RswiftCore/SwiftTypes/Type.swift [hunk 3]
```diff
@@ -60,6 +60,10 @@ struct Type: UsedTypesProvider, CustomStringConvertible, Hashable {
   static let Strings = Type(module: "Rswift", name: "Strings")
   static let Validatable = Type(module: "Rswift", name: "Validatable")
   static let TypedStoryboardSegueInfo = Type(module: "Rswift", name: "TypedStoryboardSegueInfo", genericArgs: [TypeVar(description: "Segue", usedTypes: []), TypeVar(description: "Source", usedTypes: []), TypeVar(description: "Destination", usedTypes: [])])
+  static let LocalizedStringNoArg = Type(module: "Rswift", name: "LocalizedStringNoArg")
+  static let LocalizedStringOneArg = Type(module: "Rswift", name: "LocalizedStringOneArg")
+  static let LocalizedStringTwoArgs = Type(module: "Rswift", name: "LocalizedStringTwoArgs")
+  static let LocalizedStringThreeArgs = Type(module: "Rswift", name: "LocalizedStringThreeArgs")
```
Comment (cyrilmottier) [inline Sources/RswiftCore/SwiftTypes/Type.swift]: I woudl add `LocalizedStringFourArgs` and `LocalizedStringFiveArgs` for futureproofness

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift [hunk 4]
```diff
@@ -60,8 +60,11 @@ struct StringsStructGenerator: ExternalOnlyStructGenerator {
       type: Type(module: .host, name: structName),
       implements: [],
       typealiasses: [],
-      properties: params.map { stringLet(values: $0, at: externalAccessLevel) },
-      functions: params.map { stringFunction(values: $0, at: externalAccessLevel) },
+      properties: params.flatMap {[
+        stringLet(values: $0, at: externalAccessLevel),
+        resourceLet(values: $0),
+        ]},
+      functions: [],
```
Comment (cyrilmottier) [inline Sources/RswiftCore/Generators/StringsStructGenerator.swift]: 👍 awesome

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift [hunk 5]
```diff
@@ -216,70 +219,35 @@ struct StringsStructGenerator: ExternalOnlyStructGenerator {
...
-      return Function.Parameter(name: argumentLabel, localName: valueName, type: param.spec.type)
+  private func stringLet(values: StringValues, at externalAccessLevel: AccessLevel) -> Let {
+    let type: Type
+    let argTypes = values.params.map { $0.spec.type }
+    switch argTypes.count {
+      case 0: type = .LocalizedStringNoArg
+      case 1: type = .LocalizedStringOneArg
+      case 2: type = .LocalizedStringTwoArgs
+      case 3: type = .LocalizedStringThreeArgs
+      default: fatalError("Unexpected argument count")
```
Comment (cyrilmottier) [inline Sources/RswiftCore/Generators/StringsStructGenerator.swift]: "Unhandled argument count. Maximum supported is 3"?

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift [hunk 6]
```diff
@@ -295,6 +263,16 @@ extension Locale {
   }
 }
 
+private func snakeToCamel(_ string: String) -> String {
```
Comment (cyrilmottier) [inline Sources/RswiftCore/Generators/StringsStructGenerator.swift]: When do you guys choose a method over an extension method?

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift (hunk 1 above)
Comment (scalbatty) [reply Sources/RswiftCore/Generators/StringsStructGenerator.swift]: It's a `Substring`

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift (hunk 6 above)
Comment (scalbatty) [reply Sources/RswiftCore/Generators/StringsStructGenerator.swift]: Changed it to a private extension method 👍

File: Sources/RswiftCore/SwiftTypes/Type.swift (hunk 3 above)
Comment (scalbatty) [reply Sources/RswiftCore/SwiftTypes/Type.swift]: Added them

File: Sources/RswiftCore/Generators/StringsStructGenerator.swift (hunk 2 above)
Comment (scalbatty) [reply Sources/RswiftCore/Generators/StringsStructGenerator.swift]: It's a Substring, rewriting the code for clarification

**Pull request merged**